packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts

```diff
@@ -202,51 +174,44 @@ export const useGridSelection = (apiRef: GridApiRef, props: GridComponentProps):
 
   React.useEffect(() => {
     // Rows changed
-    setGridState((state) => {
-      const newSelectionState = gridSelectionStateSelector(state);
-      const selectionLookup = selectedIdsLookupSelector(state);
-      let hasChanged = false;
-      newSelectionState.forEach((id: GridRowId) => {
-        if (!rowsLookup[id]) {
-          delete selectionLookup[id];
-          hasChanged = true;
-        }
-      });
-      if (hasChanged) {
-        return { ...state, selection: Object.values(selectionLookup) };
+    const currentSelection = gridSelectionStateSelector(apiRef.current.state);
+
+    // We clone the existing map to avoid mutating the same map returned by the selector to others part of the project
+    const selectionLookup = new Map(selectedIdsLookupSelector(apiRef.current.state).entries());
+
+    let hasChanged = false;
+    currentSelection.forEach((id: GridRowId) => {
+      if (!rowsLookup[id]) {
+        selectionLookup.delete(id);
+        hasChanged = true;
       }
-      return state;
     });
-    forceUpdate();
-  }, [rowsLookup, apiRef, setGridState, forceUpdate]);
+
+    if (hasChanged) {
+      apiRef.current.setSelectionModel([...selectionLookup.keys()]);
+    }
+  }, [rowsLookup, apiRef]);
 
   React.useEffect(() => {
     if (propSelectionModel === undefined) {
       return;
     }
 
     apiRef.current.setSelectionModel(propSelectionModel);
-  }, [apiRef, propSelectionModel, setGridState]);
+  }, [apiRef, propSelectionModel]);
 
   React.useEffect(() => {
     // isRowSelectable changed
```

Comment:
How is the state updated to the prop value if we remove this effect ?

Reply:
Even keeping this effect, the state is not updated when controlled because the prop value always win. In other rowds, the state will contain unselectable rows. This effect could be improved to only run if `selectionModel` prop is not used.

Reply:
My bad, the Github UI made me think you were talking about the effect above.
For the `isRowSelectable` effect I agree if we are in control mode, it is useless.
Do you think we should keep it if not in control mode ?

Reply:
>Do you think we should keep it if not in control mode ? 

We can keep it but only running if the selection is not controlled.

> Meaning, in non-controlled mode, do we want the state to contain non-selectable rows or not ?

To be clear, the state will only contain non-selectable rows if the `isRowSelectable` prop changes in a way where previously selected rows are now non-selectable. Besides that, the state is always correct.